notifications/src/main/kotlin/com/d3/notifications/queue/EventsQueue.kt

```diff
@@ -58,91 +62,146 @@ class EventsQueue(
      */
     fun enqueue(transferNotifyEvent: TransferNotifyEvent) {
+        enqueue(transferNotifyEvent, TRANSFERS_QUEUE_NAME)
+    }
+
+    /**
+     * Puts registration event into RabbitMQ
+     * @param registrationNotifyEvent - event to put
+     */
+    fun enqueue(registrationNotifyEvent: RegistrationNotifyEvent) {
+        enqueue(registrationNotifyEvent, REGISTRATIONS_QUEUE_NAME)
+    }
+
+    /**
+     * Puts event into RabbitMQ
+     * @param event - event to put
+     * @param queue - name of queue to put event into
+     */
+    private fun enqueue(event: Event, queue: String) {
         try {
-            logger.info("Enqueue transfer $transferNotifyEvent")
-            val json = gson.toJson(transferNotifyEvent)
+            logger.info("Enqueue event $event to queue $queue")
+            val json = gson.toJson(event)
             channel.basicPublish(
                 "",
-                TRANSFERS_QUEUE_NAME,
+                queue,
                 MessageProperties.MINIMAL_PERSISTENT_BASIC,
                 json.toByteArray()
             )
+            logger.info("Event $event has been successfully published to queue $queue")
         } catch (e: Exception) {
-            logger.error("Cannot enqueue transfer $transferNotifyEvent", e)
+            logger.error("Cannot enqueue $event", e)
         }
     }
 
     override fun close() {
-        consumerTag?.let {
-            channel.basicCancel(it)
+        consumerTags.forEach {
+            tryClose { channel.basicCancel(it) }
+        }
+        tryClose { channel.close() }
+        tryClose { connection.close() }
+        tryClose { subscriberExecutorService.shutdownNow() }
+    }
+
+    private fun tryClose(close: () -> Unit) {
+        try {
+            close()
+        } catch (e: Exception) {
+            logger.error("Cannot close", e)
         }
-        subscriberExecutorService.shutdownNow()
-        channel.close()
-        connection.close()
     }
 
     /**
      * Iterates through notification services
-     * @param transferNotifyEvent - transfer event to notify
+     * @param event - event to notify
+     * @param queueName - name of queue used for a given event
      * @param iterator - iteration logic
      */
     private fun iterateThroughNotificationServices(
-        transferNotifyEvent: TransferNotifyEvent,
+        event: Event,
+        queueName: String,
         iterator: (NotificationService) -> Result<Unit, java.lang.Exception>
     ) {
         notificationServices.forEach {
             iterator(it).failure { ex ->
                 if (ex is RepeatableException) {
                     // Re-queue if possible
-                    logger.warn("Cannot handle event due to error. Try to re-queue $transferNotifyEvent", ex)
-                    enqueue(transferNotifyEvent)
+                    logger.warn("Cannot handle event due to error. Try to re-queue $event", ex)
+                    enqueue(event, queueName)
```

Comment:
so duplicates are eliminated by that plugin, right?

Reply:
No. We have no plugins

Reply:
> No. We have no plugins

So how duplicates are prevented here

Reply:
They are not prevented.